les/clientpool.go

```diff
@@ -398,6 +475,57 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
 			return f.connectedCap > f.capLimit || f.connectedQueue.Size() > f.connLimit
 		})
 	}
+	if f.eventHook != nil {
+		f.eventHook("capacityUpdate", nil)
+	}
+}
+
+// setCapacity sets the assigned capacity of a connected client
+func (f *clientPool) setCapacity(c *clientInfo, capacity uint64) error {
+	if f.connectedMap[c.id] != c {
+		return errClientNotConnected
+	}
+	if c.capacity == capacity {
+		return nil
+	}
+	if !c.priority {
+		return errNoPriority
+	}
+	oldCapacity := c.capacity
+	c.capacity = capacity
+	f.connectedCap += capacity - oldCapacity
+	f.connectedQueue.Remove(c.queueIndex)
+	f.connectedQueue.Push(c)
+	if f.connectedCap > f.capLimit {
+		var kickList []*clientInfo
+		kick := true
+		f.connectedQueue.MultiPop(func(data interface{}, priority int64) bool {
+			client := data.(*clientInfo)
+			kickList = append(kickList, client)
+			f.connectedCap -= client.capacity
+			if client == c {
+				kick = false
+			}
+			return kick && (f.connectedCap > f.capLimit)
+		})
+		if kick {
+			now := mclock.Now()
+			for _, c := range kickList {
+				f.dropClient(c, now, true)
+			}
+		} else {
+			c.capacity = oldCapacity
+			for _, c := range kickList {
+				f.connectedCap += c.capacity
+				f.connectedQueue.Push(c)
+			}
+			return errNoPriority
+		}
+	}
+	totalConnectedGauge.Update(int64(f.connectedCap))
+	f.priorityConnected += capacity - oldCapacity
+	c.peer.updateCapacity(c.capacity)
+	return nil
```

Comment:
The problem with your version is that if `c.balanceTracker` is not updated with the new capacity and its priority is not updated in `f.connectedQueue` then `c` is still in its old place in the queue and `MultiPop` will return a wrong order. The question here is whether `c` can stay in the queue with its _new_ priority (which is affected by the capacity). So I think there is no easy way around updating and then reverting it if necessary. Also, I don't think it is more complicated, we just have to take care of properly reverting the original state if necessary before releasing the lock.

Reply:
I changed the error message to "priority too low to raise capacity" because that describes both cases perfectly. Zero priority is obviously not enough. If there is "no space" to raise capacity then that is also an issue of low priority because with a higher priority other clients could be kicked out.
I am thinking about returning the missing amount of balance in order to successfully change the capacity so that the client can decide it easier if it wants to buy the necessary amount of tokens.